File: backend/app/schemas/contract_share.py
"""Pydantic schemas for Contract Share endpoints."""

from datetime import datetime
from uuid import UUID
from typing import Optional, List

from pydantic import BaseModel, Field

from app.schemas.external_user import ExternalUserSummary


# ===== Request Schemas =====

class ContractShareCreate(BaseModel):
    """Schema for sharing a contract with an external user."""
    external_user_id: UUID
    can_download: bool = False
    can_comment: bool = True
    expires_in_days: Optional[int] = Field(None, ge=1, le=365)
    message: Optional[str] = None


class ContractShareBulkCreate(BaseModel):
    """Schema for sharing a contract with multiple external users."""
    external_user_ids: List[UUID] = Field(..., min_length=1)
    can_download: bool = False
    can_comment: bool = True
    expires_in_days: Optional[int] = Field(None, ge=1, le=365)
    message: Optional[str] = None


class ContractShareUpdate(BaseModel):
    """Schema for updating a contract share."""
    can_download: Optional[bool] = None
    can_comment: Optional[bool] = None
    expires_at: Optional[datetime] = None


# ===== Response Schemas =====

class ContractShareSummary(BaseModel):
    """Minimal contract share info."""
    id: UUID
    external_user_id: UUID
    can_download: bool
    can_comment: bool
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class ContractShareResponse(BaseModel):
    """Schema for contract share response."""
    id: UUID
    contract_id: UUID
    external_user_id: UUID
    shared_by_id: UUID
    can_download: bool
    can_comment: bool
    expires_at: Optional[datetime] = None
    message: Optional[str] = None
    access_count: int
    last_access_at: Optional[datetime] = None
    is_revoked: bool
    revoked_at: Optional[datetime] = None
    revoked_by_id: Optional[UUID] = None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ContractShareWithUser(ContractShareResponse):
    """Contract share with external user details."""
    external_user: ExternalUserSummary

    class Config:
        from_attributes = True


class ContractShareListResponse(BaseModel):
    """Schema for list of contract shares."""
    items: List[ContractShareWithUser]
    total: int


class ShareInviteResponse(BaseModel):
    """Response for share invitation with access link."""
    share: ContractShareResponse
    access_url: str
    token: str

File: frontend/src/types/contract-share.ts
// Contract Share types

export interface ContractShareCreate {
  external_user_id: string
  can_download?: boolean
  can_comment?: boolean
  expires_in_days?: number
  message?: string
}

export interface ContractShareResponse {
  id: string
  contract_id: string
  external_user_id: string
  shared_by_id: string
  can_download: boolean
  can_comment: boolean
  expires_at?: string
  message?: string
  access_count: number
  last_access_at?: string
  is_revoked: boolean
  revoked_at?: string
  created_at: string
  updated_at: string
}

export interface ExternalUserSummary {
  id: string
  email: string
  full_name?: string
  company_name?: string
}

export interface ContractShareWithUser extends ContractShareResponse {
  external_user: ExternalUserSummary
}

export interface ContractShareListResponse {
  items: ContractShareWithUser[]
  total: number
}

export interface ShareInviteResponse {
  share: ContractShareResponse
  access_url: string
  token: string
}
